### PR TITLE
Internationalized domain name (IDN) support

### LINKED ISSUE #2285: Request to cyrillic domain fails
**Guzzle version(s) affected**: 6.3.3

**Description**  

Request to cyrillic domain fails.

**How to reproduce**  

```
$httpClient = new \GuzzleHttp\Client();
$response = $httpClient->get('https://русские-домены.рф');
echo (string)$response->getBody();
```
**Possible Solution**  

It fails due to *Host* header in `GuzzleHttp\Psr7\Request::updateHostFromUri()`
Header must be converted using `idn_to_ascii()` function.

**Additional context**  

But it works using `curl`

```
$ch = curl_init('https://русские-домены.рф');
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
$html = curl_exec($ch);
curl_close($ch);
echo $html;
```

### PR DESCRIPTION
**Pull request by alexeyshockov**
This PR introduces IDN support on client's level, so all the handlers will work with the ASCII version of the IDN with no need for additional conversion.

Also fixes #2285.

### CODE REVIEW

**Comment by alexeyshockov** [comment]
The build will be green when #2278 is merged. Other tests are OK.

**Comment by alexeyshockov** [comment]
Although this PR is self-contained, guzzle/psr7#269 makes IDN support overall better.

**Comment by sagikazarmark** [comment]
@alexeyshockov Thanks! Can you please rebase your branch?

**Comment by alexeyshockov** [comment]
@sagikazarmark, done

File: src/RequestOptions.php [hunk 1]
```diff
@@ -132,6 +132,14 @@ final class RequestOptions
      */
     const HTTP_ERRORS = 'http_errors';
 
+    /**
+     * idn: (bool|int, default=true) Set to false to disable exceptions
```
Comment (hsgce) [inline src/RequestOptions.php]: Looks like this docblock needs updating.

File: tests/bootstrap.php [hunk 2]
```diff
@@ -1,4 +1,9 @@
 <?php
+
+namespace {
+    setlocale(LC_ALL, 'C');
```
Comment (hsgce) [inline tests/bootstrap.php]: Do we want all tests running in this locale?

File: src/RequestOptions.php (hunk 1 above)
Comment (alexeyshockov) [reply src/RequestOptions.php]: Thanks, fixed

**Comment by alexeyshockov** [comment]
@sagikazarmark, could you take a look also at this for the next release?

File: tests/bootstrap.php (hunk 2 above)
Comment (alexeyshockov) [reply tests/bootstrap.php]: Currently Guzzle fails on other locales. See guzzle/psr7#269 for details, it also solves this.

**Comment by sagikazarmark** [comment]
Yeah, it's on my list, thanks. I recall a related PR in the psr7 package. Is it a dependency of this PR? Do we have to raise the minimum version?

**Comment by alexeyshockov** [comment]
> Is it a dependency of this PR?

No, the PR to PSR7 is not a dependency, but an addition (very convenient, though, because without it Guzzle will struggle with locales, that's why hardcoded it to `'C'` in the tests).

> Do we have to raise the minimum version?

Minimum version of PHP stays the same (IDN support is enabled only when `intl` extension is available).

**Comment by sagikazarmark** [comment]
> Minimum version of PHP stays the same (IDN support is enabled only when intl extension is available).

I meant PSR7 package version, but it's not necessary then.

**Comment by alexeyshockov** [comment]
Yeah, I mixed things with PSR7 package a bit, will prepare a separate PR to `1.x` branch

**Comment by Nyholm** [comment]
I think this looks good. Could you add a test where you use base_uri?

**Comment by Nyholm** [comment]
If you rebase this on master and then add a test where you use an internationalized domain in `base_uri`, then I would be happy to merge.

**Comment by alexeyshockov** [comment]
@Nyholm, done

**Pull request merged**